cmd/hydrate/src/lib.rs

```diff
@@ -0,0 +1,217 @@
+            bail!("region is not large enough; {end:#x} > {:#x}", mem.len());
+        }
+        data.copy_from_slice(&mem[offset..end]);
+        Ok(())
+    }
+}
+
+fn run(context: &mut ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = Args::try_parse_from(subargs)?;
+    let f = std::fs::File::open(&subargs.file)?;
+    let mut z = zip::ZipArchive::new(f)?;
+
+    let mut s = String::new();
+    z.by_name("meta.json")?.read_to_string(&mut s)?;
+    #[derive(serde::Deserialize)]
+    struct Meta {
+        version: u64,
+    }
+    let meta: Meta = serde_json::from_str(&s)?;
+    if meta.version != 1 {
+        bail!(
+            "invalid version in `meta.json`: expected 1, got {}",
+            meta.version
+        );
+    }
+
+    let mut s = String::new();
+    z.by_name("TASK_INDEX")?.read_to_string(&mut s)?;
+    let task_id: u16 = s.trim().parse()?;
+
+    let mut s = String::new();
+    z.by_name("TIMESTAMP")?.read_to_string(&mut s)?;
+    let timestamp: u64 = s.trim().parse()?;
+
+    let mut bord = String::new();
+    z.by_name("BORD")?.read_to_string(&mut bord)?;
+    bord.pop(); // remove newline
+
+    let mut gitc = String::new();
+    z.by_name("GITC")?.read_to_string(&mut gitc)?;
+    gitc.pop(); // remove newline
+
+    let vers = match z.by_name("VERS") {
+        Ok(mut v) => {
+            let mut s = String::new();
+            v.read_to_string(&mut s)?;
+            s.pop();
+            Some(s)
+        }
+        Err(zip::result::ZipError::FileNotFound) => None,
+        Err(e) => return Err(e.into()),
+    };
+
+    let mut s = String::new();
+    z.by_name("ARCHIVE_ID")?.read_to_string(&mut s)?;
+    let mut archive_id = [0u8; 8];
+    for (i, b) in archive_id.iter_mut().enumerate() {
+        *b = u8::from_str_radix(&s[i * 2..][..2], 16)?;
+    }
+
+    let mut mem = BTreeMap::new();
+    let mem_files = z
+        .file_names()
+        .filter(|f| f.starts_with("0x"))
+        .map(|s| s.to_owned())
+        .collect::<Vec<_>>();
+    for f in mem_files {
+        let num = f.strip_suffix(".bin").unwrap().strip_prefix("0x").unwrap();
+        let addr = u32::from_str_radix(num, 16)?;
+        let mut data = z.by_name(&f)?;
+        let mut v = vec![];
+        data.read_to_end(&mut v)?;
+        mem.insert(addr, v);
+    }
+
+    msg!("read dehydrated crash dump");
+    msg!("  task index: {task_id}");
+    msg!("  crash time: {timestamp}");
+    msg!("  archive id: {archive_id:02x?}");
+    msg!("  board:      {bord}");
+    msg!("  git commit: {gitc}");
+    msg!("  version:    {}", vers.as_deref().unwrap_or("[missing]"));
+    msg!("  {} memory regions:", mem.len());
+    for (k, v) in &mem {
+        msg!("    {k:#08x}: {} bytes", v.len());
+    }
+
+    // compare archive ID
+    if context.cli.archive.is_none() {
+        bail!("must provide an archive");
+    }
+    let archive = context.archive.as_ref().unwrap();
```

Comment:
nitpicky, feel free to ignore: why not golf these two down to just:
```suggestion
    let Some(archive) = context.archive.as_ref() else {
        bail!("must provide an archive");
    };
```

Reply:
That's not _exactly_ equivalent, because `context.cli.archive` and `context.archive` are different things!  However, I'm not sure why they'd ever differ, so this is probably a fine suggestion; done

Reply:
Oh! I can't read --- I thought they were both `context.archive`. Whoops!